### PR TITLE
Attempt to test against jormungandr v0.3.1 in CI

### PR DESCRIPTION
**Pull request by piotr-iohk**
# Issue Number

# Overview

- [ ] I have changed link to jormungandr@0.3.1 in `travis.yaml` and updated config generation

# Comments

I'm pretty sure that more stuff need to be updated.

### CODE REVIEW

**Comment by KtorZ** [comment]
Turns out that this is going to be slightly trickier than expected. Since `v0.3.0`, Jörmungandr changed the way it computes transaction ids (hence the many failures for scenarios that involves making transactions). 

Now, the id is made of the full transaction content (including witnesses), which means that when de-serializing transaction we need to keep corresponding witnesses close by and store them alongside the transactions we discover. 

Fortunately, our wallet core engine is already parametrized by the transaction type, but we need some fiddling with the database to make sure we can serialize and unserialize tx for both backends... (instead of the hacky hack we currently have :s ... )

**Comment by KtorZ** [comment]
An _rather up-to-date_ documentation of this new format can be found here: https://github.com/input-output-hk/chain-libs/blob/self-cert/chain-impl-mockchain/doc/format.md

File: lib/core/src/Cardano/Wallet/DB/Sqlite.hs [hunk 1]
```diff
@@ -602,9 +602,12 @@ txHistoryFromEntity
...
 txHistoryFromEntity metas ins outs = map mkItem metas
   where
-    mkItem m = ( getTxId (txMetaTxId m)
-               , (mkTxWith (txMetaTxId m), mkTxMeta m) )
+    mkItem m =
+        ( getTxId (txMetaTxId m)
+        , (mkTxWith (txMetaTxId m), mkTxMeta m)
+        )
     mkTxWith txid = mkTx @t
+        (getTxId txid)
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/DB/Sqlite.hs:610]: :+1:

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs [hunk 2]
```diff
@@ -228,48 +237,67 @@ data Message
     -- Messages not yet supported go there.
     deriving (Eq, Show)
 
+data MessageType
+    = MsgTypeInitial
+    | MsgTypeTransaction
+
+data TxWitnessTag
+    = TxWitnessLegacyUTxO
+    | TxWitnessUTxO
```
Comment (paweljakubas) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:246]: general question - what is rationale to have both `TxWitnessLegacyUTxO` and `TxWitnessUTxO` - maybe when used will be evident...

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs [hunk 3]
```diff
@@ -530,12 +553,65 @@ isolatePut l x = do
...
+--     SIZE (2 bytes) | TYPE (1 byte)
+--
+withHeader :: MessageType -> Put -> Put
+withHeader typ content = do
+    let bs = BL.toStrict $ runPut (putWord8 (messageTypeTag typ) *> content)
     putWord16be (toEnum $ BS.length bs)
     putByteString bs
 
+-- | Compute a Blake2b_256 hash of a given 'ByteString'
+blake2b256 :: ByteString -> ByteString
```
Comment (paweljakubas) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:571]: maybe use `blake2b256` from `lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs` ?

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Primitive/Types.hs [hunk 4]
```diff
@@ -28,7 +30,13 @@ import GHC.Generics
     ( Generic )
 
 data Tx = Tx
-    { inputs
+    { txid
+        :: Hash "Tx"
+        -- ^ Jörmungandr computes transaction id by hashing the full content of
+        -- the transaction, which includes witnesses. Therefore, we need either
+        -- to keep track of the witnesses to be able to re-compute the tx id
+        -- every time, or, simply keep track of the id itself.
```
Comment (paweljakubas) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Primitive/Types.hs:38]: hope in the next version of Jormungandr will not change how  they compute transaction id again ...

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs [hunk 5]
```diff
@@ -57,17 +58,29 @@ newTransactionLayer
...
-        let tx = Tx (fmap (second coin) inps) outs
-        let bs = block0 <> getHash (txId @(Jormungandr n) tx)
-        txWitnesses <- forM inps $ \(_, TxOut addr _) -> sign bs
+    { mkStdTx = \keyFrom rnps outs -> do
+        -- NOTE
+        --
+        -- Yet, for signing, we need to embed a hash of the transaction data
+        -- without the witnesses (since we don't yet have them!). In this sense,
+        -- this is a transaction id as Byron nodes or the http-bridge
+        -- defines them.
```
Comment (paweljakubas) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs:66]: thanks for this comment!

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs [hunk 6]
```diff
@@ -57,17 +58,29 @@ newTransactionLayer
...
 
-    -- FIXME:
-    -- Implement fee calculation for Jörmungandr!
-    -- See: https://github.com/input-output-hk/cardano-wallet/blob/f683a6d609bed3bea02eca1a18205d84f6486bd6/lib/jormungandr/test/integration/Main.hs#L217-L237
-    , estimateSize = \_ -> Quantity 0
+    -- NOTE
+    -- Jörmungandr fee calculation is a linear function where the coefficient
+    -- is multiplied by the total number of inputs and outputs.
+    , estimateSize = \(CoinSelection inps outs chgs) ->
+        Quantity $ length inps + length outs + length chgs
```
Comment (paweljakubas) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs:82]: :+1:

**Review by paweljakubas — APPROVED** [review]
Nice work! Hope next versions of Jormungandr will not require such an effort!

**Comment by paweljakubas** [comment]
one more remark : maybe it is good idea to include steps to install Jormungandr to https://github.com/input-output-hk/cardano-wallet/wiki/Testing - we have here how to install cardano-node, http-bridge but Jormungandr is missing. what do you think?

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs (hunk 2 above)
Comment (KtorZ) [reply lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:246]: This is a direct translation of the Rust's model. Jörmungandr has a new format for signing transaction. Though, for compatibility reason, for legacy addresses and transaction from legacy addresses, Jörmungandr still allow the old format (with the CBOR tags) to be used. So, this is how we tell Jörmungandr about which format is being used.

(note however that this was there before this update, nothing is new here).

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs (hunk 3 above)
Comment (KtorZ) [reply lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:571]: Hmmm... I'd rather move this function in a common place that is neither Jormungandr.Binary, nor AddressDerivation.Random ^^" ... But good point.

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Primitive/Types.hs (hunk 4 above)
Comment (KtorZ) [reply lib/jormungandr/src/Cardano/Wallet/Jormungandr/Primitive/Types.hs:38]: I think that decision is rather final (they basically now compute only "fragment ids", so the idea is even more holistic than a transaction id. A transaction is one type of "fragment", but there are many other types, and it allows for an open extension). 

Yet, they might change some other stuff that may not be trivial to update either :man_shrugging:

**Comment by KtorZ** [comment]
> maybe it is good idea to include steps to install Jormungandr

Yes and no. We've included instructions for http-bridge because we maintain a custom fork and, because the development of that component is mostly _done_. 
However, for Jörmungandr, I'd rather point to the repository and leave it to the Rust team to keep their documentation up-do-date about how to install the software :)

**Pull request merged**